cmd/xl-v1.go

```diff
@@ -195,45 +201,222 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+	var disks []StorageAPI
+	buckets, err := xl.ListBuckets(ctx)
+	if err != nil {
+		return err
+	}
 	for _, d := range xl.getLoadBalancedDisks() {
 		if d == nil || !d.IsOnline() {
 			continue
 		}
-		randomDisks = append(randomDisks, d)
-		if len(randomDisks) >= 3 {
-			break
+		disks = append(disks, d)
+	}
+	if len(disks) == 0 || len(buckets) == 0 {
+		return nil
+	}
+
+	// Load bucket totals
+	oldCache := dataUsageCache{}
+	var buf bytes.Buffer
+	err = xl.GetObject(ctx, minioMetaBackgroundOpsBucket, dataUsageCacheName, 0, -1, &buf, "", ObjectOptions{})
+	if err != nil {
+		if !isErrObjectNotFound(err) {
+			return toObjectErr(err, minioMetaBackgroundOpsBucket, dataUsageCacheName)
+		}
+	} else {
+		err := oldCache.deserialize(buf.Bytes())
+		if err != nil {
+			logger.LogIf(ctx, err)
+			oldCache = dataUsageCache{}
+		}
+	}
+
+	// New cache..
+	cache := dataUsageCache{
+		Info: dataUsageCacheInfo{
+			Name:      dataUsageRoot,
+			NextCycle: oldCache.Info.NextCycle,
+		},
+		Cache: make(map[dataUsageHash]dataUsageEntry, len(oldCache.Cache)),
+	}
+
+	// Put all buckets into channel.
+	bucketCh := make(chan BucketInfo, len(buckets))
+	// Add new buckets first
+	for _, b := range buckets {
+		if oldCache.find(b.Name) == nil {
+			bucketCh <- b
+		}
+	}
+	// Add existing buckets.
+	for _, b := range buckets {
+		e := oldCache.find(b.Name)
+		if e != nil {
+			bucketCh <- b
+			cache.replace(b.Name, dataUsageRoot, *e)
 		}
 	}
 
-	var dataUsageResults = make([]DataUsageInfo, len(randomDisks))
+	close(bucketCh)
+	buckets = nil
+	bucketResults := make(chan dataUsageEntryInfo, len(disks))
 
+	save := func(cache *dataUsageCache, name string) error {
+		b := cache.serialize()
+		size := int64(len(b))
+		r, err := hash.NewReader(bytes.NewReader(b), size, "", "", size, false)
+		if err != nil {
+			return err
+		}
+
+		_, err = xl.PutObject(ctx,
+			minioMetaBackgroundOpsBucket,
+			name,
+			NewPutObjReader(r, nil, nil),
+			ObjectOptions{})
+		return err
+	}
+
+	// Start async collector/saver.
+	// This goroutine owns the cache.
+	var saverWg sync.WaitGroup
+	saverWg.Add(1)
+	go func() {
+		const updateTime = 30 * time.Second
+		t := time.NewTicker(updateTime)
+		defer t.Stop()
+		defer saverWg.Done()
+		var lastSave time.Time
+
+	saveLoop:
+		for {
+			select {
+			case <-ctx.Done():
+				// Return without saving.
+				return
+			case <-t.C:
+				if cache.Info.LastUpdate.Equal(lastSave) {
+					continue
+				}
+				err := save(&cache, dataUsageCacheName)
+				updates <- cache.clone()
+				logger.LogIf(ctx, err)
+				lastSave = cache.Info.LastUpdate
+			case v, ok := <-bucketResults:
+				if !ok {
+					break saveLoop
+				}
+				cache.replace(v.Name, v.Parent, v.Entry)
+				cache.Info.LastUpdate = time.Now()
+			}
+		}
+		// Save final state...
+		cache.Info.NextCycle++
+		cache.Info.LastUpdate = time.Now()
+		logger.LogIf(ctx, save(&cache, dataUsageCacheName))
+		updates <- cache
+	}()
+
+	// Start one crawler per disk
```

Comment:
Are we sure we want to crawl all disks here ?

Reply:
Yes. We aim to spread the load as much as possible.